### PR TITLE
Refactor and add Testing

### PR DESCRIPTION
**Pull request by ahmadnassri**
## Description

Refactoring and add Testing

## Motivation and Context

I hear testing is a thing 

## How Has This Been Tested?

well the thing is ... this adds testing,

## Types of changes

- [ ] Bugfix
- [ ] Feature
- [ ] Performance Improvement
- [ ] Tests _(fix, improvement, new tests)_
- [ ] Code style update _(formatting, local variables)_
- [x] Refactoring _(no functional changes, no api changes)_
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other... Please describe: 

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have run `npm test` locally and all tests are passing.
- [x] I have added/updated tests for any new behavior.
- [x] I have reviewed the [guidelines for contributing](../CONTRIBUTING.md) to this repository.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,3 +3,5 @@
 /.scripts/cc-test-reporter
 /coverage
 /node_modules
+/example
+.DS_Store
```
Comment (ahmadnassri) [inline .gitignore]: don't add non-project ignore files, use your personal global config under your user directory `~/.gitignore`

File: .gitignore [hunk 2]
```diff
@@ -3,3 +3,5 @@
 /.scripts/cc-test-reporter
 /coverage
 /node_modules
+/example
```
Comment (ahmadnassri) [inline .gitignore]: what is `/example` ?

File: lib/events.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
+const { promisify } = require('util')
+const { gunzip } = require('zlib')
+const { send } = require('./utility')
+const redis = require('./redis')
+const { debug } = require('./logger')
+const decompress = promisify(gunzip)
```
Comment (ahmadnassri) [inline lib/events.js]: as a nice to do, organize dependencies per module alphabetically and separate external vs. internal with line breaks:

```js
const { gunzip } = require('zlib')

const { debug } = require('./logger')
const { promisify } = require('util')
const { send } = require('./utility')
const redis = require('./redis')

const decompress = promisify(gunzip)
```

File: package.json [hunk 4]
```diff
@@ -33,10 +33,12 @@
     "pretest": "npm run test:codeclimate",
     "test": "npm run test:tap",
     "test:codeclimate": ".scripts/codeclimate analyze",
-    "test:tap": "tap test --100 --coverage-report=lcov --no-browser",
+    "test:tap": "tap test --coverage-report=lcov --no-browser",
```
Comment (ahmadnassri) [inline package.json]: keep the 100 as a reminder ;)

File: test/events.spec.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
+
```
Comment (ahmadnassri) [inline test/events.spec.js]: `.spec.` suffix is not needed ... it's a relic of other opinionated tooling ...

File: test/events.spec.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    }
+    const res = {
+      statusCode: 200,
+      statusMessage: 'OK',
+      headers: {
+
+      },
+      on: sandbox.stub()
+    }
+    const data = JSON.stringify({ "page": 1, "per_page": 3, "total": 12, "total_pages": 4, "data": [{ "id": 1, "first_name": "George", "last_name": "Bluth", "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/calebogden/128.jpg" }, { "id": 2, "first_name": "Janet", "last_name": "Weaver", "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/josephstein/128.jpg" }, { "id": 3, "first_name": "Emma", "last_name": "Wong", "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/olegpogodaev/128.jpg" }] })
```
Comment (ahmadnassri) [inline test/events.spec.js]: store your fixtures in static files under `test/fixtures/`

File: test/handler.spec.js [hunk 7]
```diff
@@ -0,0 +1,152 @@
+const tap = require('tap')
+const followRedirects = require('follow-redirects')
+const sinon = require('sinon')
+const handler = require('../lib/handler')
+const redisClient = require('../lib/redis')
+const events = require('../lib/events')
+
+const sampleResponse = {
```
Comment (ahmadnassri) [inline test/handler.spec.js]: see note about fixtures above

File: test/handler.spec.js [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+  end()
+})
+
+tap.afterEach((next) => {
+  sandbox.restore()
+  next()
+})
+
+tap.test('handler', (t) => {
```
Comment (ahmadnassri) [inline test/handler.spec.js]: you don't need to wrap all the tests with a top level one.

you can simply call multiple `tap.test` on the root of the module

File: lib/logger.js [hunk 9]
```diff
@@ -0,0 +1,4 @@
+const { debuglog } = require('util')
+const debug = debuglog('cachepile')
```
Comment (ahmadnassri) [inline lib/logger.js]: just export directly, no need for extra variables.

```js
const { debuglog } = require('util')
module.exports = debuglog('cachepile')
```

File: lib/handler.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
+const http = require('follow-redirects').http
+const https = require('follow-redirects').https
```
Comment (ahmadnassri) [inline lib/handler.js]: same dependency, use destructing in one call

File: .gitignore (hunk 2 above)
Comment (alfredctchoi) [reply .gitignore]: example is the sample project I use for local testing.

File: package.json (hunk 4 above)
Comment (alfredctchoi) [reply package.json]: I thought I put it back in 😞

**Pull request merged**